### PR TITLE
Make cork's --sdk-version optional

### PR DESCRIPTION
**Pull request by bassamtabbara**
cork create --sdk-version is now optional. If not specified we will read COREOS_SDK_VERSION from version.txt in the manifest repo. 

I'm shelling out to "git clone" instead of using the git golang lib, hope thats OK.

### CODE REVIEW

File: sdk/download.go [hunk 1]
```diff
@@ -318,3 +319,18 @@ func UpdateSignedFile(file, url string, client *http.Client) error {
...
 }
+
+func DetectSDKVersion(url, branch string) (string, error) {
+	tmp, err := ioutil.TempDir("", "")
+	if err != nil {
+		return "", err
+	}
+	defer os.RemoveAll(tmp)
+
+	clone := exec.Command("git", "clone", "--depth=1", "--single-branch", "-b", branch, url, tmp)
```
Comment (marineam) [inline sdk/download.go]: At the very least connect stderr so a failure can be diagnosed.

Comment (bassamtabbara) [inline sdk/download.go]: Ok. its noisy though which is why I removed it :-)

Comment (bassamtabbara) [inline sdk/download.go]: This is what it looks like:

```
2015-12-28T23:16:40Z cork: Detecting SDK version
Cloning into '/tmp/732493167'...
remote: Counting objects: 9, done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 9 (delta 0), reused 5 (delta 0), pack-reused 0
Unpacking objects: 100% (9/9), done.
Checking connectivity... done.
```

File: sdk/version.go [hunk 2]
```diff
@@ -53,8 +51,24 @@ func GetVersion(dir string) (ver string, err error) {
 	return
 }
 
+func GetVersion() (string, error) {
```
Comment (marineam) [inline sdk/version.go]: Remove, this is just duplicating `GetLatestVersion` which seems like the better name anyway.

File: sdk/version.go [hunk 3]
```diff
@@ -53,8 +51,24 @@ func GetVersion(dir string) (ver string, err error) {
...
+	return getVersion(dir, "COREOS_VERSION=")
+}
+
 func GetLatestVersion() (string, error) {
-	return GetVersion(BuildImageDir("latest"))
+	return getVersion(BuildImageDir("latest"), "COREOS_VERSION=")
+}
+
+func GetSDKVersion() (string, error) {
+	return getVersion(BuildImageDir("latest"), "COREOS_SDK_VERSION=")
```
Comment (marineam) [inline sdk/version.go]: I'm not sure this would ever be useful, but it would be useful to have a shorthand for getting the SDK version from `.repo/manifests`, perhaps change this to do that instead.

File: sdk/download.go (hunk 1 above)
Comment (marineam) [inline sdk/download.go]: Ah, and git doesn't have a quiet option? Perhaps then use CombinedOutput and dump on error.

Comment (bassamtabbara) [inline sdk/download.go]: ok it looks like passing -q to git clone will avoid the extra noise:

```
 cork create --replace
2015-12-28T23:21:09Z cork: Detecting SDK version
2015-12-28T23:21:10Z cork: Downloading SDK version 891.0.0
```

and here it is when something is wrong

```
 cork create --replace --manifest-url=foo
2015-12-28T23:20:08Z cork: Detecting SDK version
fatal: repository 'foo' does not exist
2015-12-28T23:20:08Z cork: DetectSDKVersion failed: exit status 128
```

File: sdk/download.go [hunk 4]
```diff
@@ -318,3 +319,18 @@ func UpdateSignedFile(file, url string, client *http.Client) error {
 	plog.Infof("Verified file: %s", file)
 	return nil
 }
+
+func DetectSDKVersion(url, branch string) (string, error) {
```
Comment (marineam) [inline sdk/download.go]: After https://github.com/coreos/mantle/pull/220/files#r48512571 you can call GetSDKVersion to re-use an existing repo clone instead of fetching.

File: sdk/download.go (hunk 1 above)
Comment (marineam) [inline sdk/download.go]: You can also do things like append -q unless the log level is INFO or DEBUG

File: sdk/download.go (hunk 4 above)
Comment (bassamtabbara) [inline sdk/download.go]: are you saying the git clone can be avoided to get the SDK version?

**Comment by bassamtabbara** [comment]
ok I think I incorporated all your feedback and squashed the commits.

File: sdk/version.go [hunk 5]
```diff
@@ -53,8 +51,20 @@ func GetVersion(dir string) (ver string, err error) {
...
+	return getVersion(dir, "COREOS_VERSION=")
+}
+
 func GetLatestVersion() (string, error) {
-	return GetVersion(BuildImageDir("latest"))
+	return getVersion(BuildImageDir("latest"), "COREOS_VERSION=")
+}
+
+func GetSDKVersion() (string, error) {
+	return getVersion(filepath.Join(RepoRoot(), "manifests"), "COREOS_SDK_VERSION=")
```
Comment (marineam) [inline sdk/version.go]: Repo root points to the top level of the whole tree, so this should be `Join(RepoRoot(), ".repo", "manifests")`

File: sdk/download.go (hunk 4 above)
Comment (marineam) [inline sdk/download.go]: Still need to add the `GetSDKVersion` to avoid the git clone if the repo tree is already there. Most commonly this would be the case if `create --replace` is used.

Comment (bassamtabbara) [inline sdk/download.go]: ``create --replace``` could change the manifest branch and url, so we would need to handle that case too right?

Comment (marineam) [inline sdk/download.go]: Yes, but not sure what I want that behaviour to be. Maybe replace could start skipping the repo init/sync if it isn't needed, or at least make it optional but I think picking that behavior is orthogonal to this.

Comment (bassamtabbara) [inline sdk/download.go]: maybe I'm missing something here but I thought ```cork create``` (with or without replace) always called ```repo init``` which could retrieve a new manifest. if we get the ```COREOS_SDK_VERSION``` from .repo/manifests we could end up with the wrong version, right?

Comment (marineam) [inline sdk/download.go]: It may be behind but the subsequent update_chroot and/or build_packages will upgrade the chroot to (hopefully) the same state either way. The update_chroot command always happens as part of the create process in the old cros_sdk implementation, probably will be adding it to cork at some point too. If it is a concern maybe a `--sdk-version=check` or `git` or similar keyword could be used to always perform the git clone.

Basically my thought is to use locally cached stuff when reasonable by default, that way there are fewer things to tweak or fix when I find myself working remotely with limited or no internet access and still want my system to behave in a sane way.

Comment (bassamtabbara) [inline sdk/download.go]: Fair. I'll check the local repo first for the version and if that fails, the remote repo for now. Once the replace and repo init interactions are figured out we can change the SDK detection behavior too. 

FWIW, it would be good if cork create --replace replaced everything in the chroot to match the new branch and manifest. Maybe even use  --force-sync when doing repo sync.

**Comment by marineam** [comment]
LGTM, thanks!
